In [ ]:
from mini_latent_pd.models.minifold.model.model import MiniFoldModel

In [1]:
import tree

In [3]:
import torch
import torch.nn.functional as F
from pathlib import Path
import h5py
from mini_latent_pd.models.minifold.model.model import MiniFoldModel
from mini_latent_pd.models.minifold.data.config import model_config
from mini_latent_pd.models.minifold.data.of_data import of_inference
from mini_latent_pd.models.minifold.utils.residue_constants import restype_order_with_x_inverse
from esm.pretrained import load_model_and_alphabet

def load_minifold_model(cache_dir="./minifold_cache", model_size="48L", device="cuda"):
    """
    Loads the MiniFold model, alphabet, and config for interactive use.
    """
    # 1. Paths
    cache = Path(cache_dir).expanduser()
    checkpoint_path = cache / f"minifold_{model_size}.ckpt"
    
    if not checkpoint_path.exists():
        raise FileNotFoundError(f"Checkpoint not found at {checkpoint_path}. Run predict.py once to download it.")

    print(f"Loading checkpoint from {checkpoint_path}...")
    
    # 2. Load Checkpoint & Hyperparams
    ckpt = torch.load(checkpoint_path, map_location="cpu")
    hparams = ckpt["hyper_parameters"]
    
    # 3. Initialize Model
    # We create the config required for feature generation
    config_of = model_config(
        "initial_training",
        train=False,
        low_prec=False,
        long_sequence_inference=False,
    )
    
    model = MiniFoldModel(
        esm_model_name=hparams["esm_model_name"],
        num_blocks=hparams["num_blocks"],
        no_bins=hparams["no_bins"],
        config_of=config_of,
        use_structure_module=True,
        kernels=False, # Important: False for Mac/MPS compatibility
    )

    # 4. Load State Dict
    state_dict = ckpt["state_dict"]
    # Clean keys as done in predict.py
    state_dict = {k: v for k, v in state_dict.items() if "boundaries" not in k}
    state_dict = {k: v for k, v in state_dict.items() if "mid_points" not in k}
    state_dict = {k.replace("_orig_mod.", ""): v for k, v in state_dict.items()}
    state_dict = {k.replace("model.", ""): v for k, v in state_dict.items()}
    
    model.load_state_dict(state_dict, strict=False)
    model.to(device)
    model.eval()
    
    # 5. Load Alphabet (needed for tokenizing sequences)
    _, alphabet = load_model_and_alphabet(hparams["esm_model_name"])
    
    return model, alphabet, config_of.data

In [2]:
class MinifoldCollator:
    def __init__(self, alphabet, config):
        self.alphabet = alphabet
        self.config = config

    def __call__(self, batch):
        """
        Transforms a list of mdCATH dicts into a MiniFold batch.
        """
        seq_tokens_list = []
        masks_list = []
        batch_of_list = []
        gt_coords_list = []
        ids = []

        for item in batch:
            seq_str = item["sequence"]
            ids.append(item["id"])
            
            # --- A. Generate Features ---
            # This replicates predict.py's `prepare_input`
            of_feats = of_inference(seq_str, "predict", self.config)
            
            # Reconstruct clean sequence from features to ensure alignment
            clean_seq = "".join(
                [restype_order_with_x_inverse[x.item()] for x in of_feats["aatype"]]
            )[: of_feats["seq_length"]]
            
            # Encode for ESM
            encoded_seq = torch.tensor(self.alphabet.encode(clean_seq), dtype=torch.long)
            
            # --- B. Collect ---
            seq_tokens_list.append(encoded_seq)
            masks_list.append(of_feats["seq_mask"][:, 0].bool())
            
            # Filter OpenFold features to only what MiniFold needs
            relevant = {"aatype", "seq_mask", "residx_atom37_to_atom14", "atom37_atom_exists"}
            batch_of_list.append({k: v for k, v in of_feats.items() if k in relevant})
            
            # Collect Ground Truth Coords (from mdCATH)
            gt_coords_list.append(torch.tensor(item["coords"]))

        # --- C. Padding ---
        max_len = max(len(s) for s in seq_tokens_list)
        
        # 1. Pad ESM Tokens (1 is usually padding in ESM alphabet)
        padded_seqs = torch.stack([
            F.pad(s, (0, max_len - len(s)), value=1) for s in seq_tokens_list
        ])
        
        # 2. Pad Masks
        padded_masks = torch.stack([
            F.pad(m, (0, max_len - len(m)), value=0) for m in masks_list
        ])
        
        # 3. Pad OpenFold Features
        batched_of = {}
        if batch_of_list:
            for k in batch_of_list[0].keys():
                feats = [d[k] for d in batch_of_list]
                # Pad the first dimension (sequence length)
                batched_of[k] = torch.stack([
                    F.pad(f, [0] * 2 * (len(f.shape) - 1) + [0, max_len - f.shape[0]], value=0) 
                    for f in feats
                ])

        # 4. Pad Ground Truth Coords
        # shape: (Batch, Max_Len, 3)
        padded_coords = torch.nn.utils.rnn.pad_sequence(
            gt_coords_list, batch_first=True, padding_value=0.0
        )

        return {
            "id": ids,
            "seq": padded_seqs,
            "mask": padded_masks,
            "batch_of": batched_of,
            "gt_coords": padded_coords
        }

In [3]:
import torch
from datasets import IterableDataset
from torch.utils.data import DataLoader
from torch.nn.utils.rnn import pad_sequence

# Import the generator from your new file
# (Ensure mdcath_dataset.py is in the same folder as your notebook)
from mini_latent_pd.data.mdcath_dataset import mdcath_generator

# --- Configuration ---
REPO_ID = "compsciencelab/mdCATH"
SUB_SAMPLE_PROTEINS = 4 
BATCH_SIZE = 1

# --- Padding Collator (Keep this in notebook as it's simple) ---
def padding_collate(batch):
    coords_list = [torch.tensor(item["coords"]) for item in batch]
    coords_padded = pad_sequence(coords_list, batch_first=True, padding_value=0.0)
    lengths = torch.tensor([len(c) for c in coords_list])
    B, Max_L = coords_padded.shape[:2]
    mask = torch.arange(Max_L).expand(B, Max_L) < lengths.unsqueeze(1)

    return {
        "id": [item["id"] for item in batch],
        "coords": coords_padded,
        "mask": mask,
        "temp": torch.tensor([item["temp"] for item in batch]),
        "lengths": lengths
    }

# --- Execution ---

if __name__ == "__main__":
    print("--- 1. Initializing Streaming Dataset ---")
    
    # Now this works because mdcath_generator is an imported function
    ds = IterableDataset.from_generator(
        mdcath_generator, 
        gen_kwargs={"repo_id": REPO_ID, "sub_sample_proteins": SUB_SAMPLE_PROTEINS}
    )
    
    # Optional: Hugging Face dataset mapping
    ds = ds.map(lambda x: {"num_atoms": len(x["coords"])})

    print("--- 2. Running DataLoader with Padding ---")
    dataloader = DataLoader(ds, batch_size=BATCH_SIZE, collate_fn=padding_collate)

    for i, batch in enumerate(dataloader):
        print(f"\nBatch {i+1}")
        print(f"  IDs: {batch['id']}")
        print(f"  Coords Shape: {batch['coords'].shape} (Padded)")
        
        if i >= 4: 
            print("Test complete.")
            break

/Users/moustholmes/mini_latent_pd/.venv/lib/python3.12/site-packages/tqdm/auto.py:21: TqdmWarning: IProgress not found. Please update jupyter and ipywidgets. See https://ipywidgets.readthedocs.io/en/stable/user_install.html
  from .autonotebook import tqdm as notebook_tqdm


--- 1. Initializing Streaming Dataset ---
--- 2. Running DataLoader with Padding ---
Generator finding files... (using subset of 4)

Batch 1
  IDs: ['2cw7A03']
  Coords Shape: torch.Size([1, 2769, 3]) (Padded)

Batch 2
  IDs: ['2cw7A03']
  Coords Shape: torch.Size([1, 2769, 3]) (Padded)

Batch 3
  IDs: ['2cw7A03']
  Coords Shape: torch.Size([1, 2769, 3]) (Padded)

Batch 4
  IDs: ['2cw7A03']
  Coords Shape: torch.Size([1, 2769, 3]) (Padded)

Batch 5
  IDs: ['2cw7A03']
  Coords Shape: torch.Size([1, 2769, 3]) (Padded)
Test complete.


In [1]:
import urllib.request
from pathlib import Path

# Config
cache_dir = Path("./minifold_cache")
model_size = "12L" # or "48L" if you want the larger one
filename = f"minifold_{model_size}.ckpt"
target_path = cache_dir / filename

# Create dir
cache_dir.mkdir(parents=True, exist_ok=True)

# Download if missing
if not target_path.exists():
    print(f"Downloading {filename}...")
    url = f"https://huggingface.co/jwohlwend/minifold/resolve/main/minifold_{model_size}_final.ckpt"
    urllib.request.urlretrieve(url, str(target_path))
    print("Download complete.")
else:
    print(f"Found checkpoint at {target_path}")

Download complete.


In [4]:
def crop_batch(batch, max_len=600):
    """
    Hard-crops a batch to a maximum length to prevent OOM/Buffer errors.
    """
    current_len = batch["seq"].shape[1]
    if current_len <= max_len:
        return batch
    
    print(f"⚠️ Cropping batch from {current_len} to {max_len} for testing...")
    
    # Crop sequence and mask
    new_batch = {
        "seq": batch["seq"][:, :max_len],
        "mask": batch["mask"][:, :max_len],
        "batch_of": {},
        "id": batch["id"] # Keep IDs
    }
    
    # Crop OpenFold features (careful with dimensions)
    for k, v in batch["batch_of"].items():
        if v.shape[1] == current_len:
             # Assuming dim 1 is sequence length (Batch, Len, ...)
            new_batch["batch_of"][k] = v[:, :max_len, ...]
        else:
            # Some features might not map directly to len (check specific keys if needed)
            new_batch["batch_of"][k] = v
            
    return new_batch

In [11]:
from mini_latent_pd.data.mdcath_dataset import mdcath_generator

# 1. Setup Device
device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
print(f"Using device: {device}")

# 2. Load Model
# Ensure you point cache_dir to where predict.py downloaded weights
model, alphabet, config = load_minifold_model(cache_dir="/Users/moustholmes/mini_latent_pd/checkpoints/minifold_cache", model_size="12L", device=device)

# 3. Prepare Dataset
# (Assuming mdcath_generator is defined in a previous cell)
from torch.utils.data import DataLoader
from datasets import IterableDataset

ds = IterableDataset.from_generator(
        mdcath_generator, 
        gen_kwargs={"repo_id": REPO_ID, "sub_sample_proteins": SUB_SAMPLE_PROTEINS}
    )
collator = MinifoldCollator(alphabet, config)

# 4. Create DataLoader
loader = DataLoader(ds, batch_size=2, collate_fn=collator)

# 5. Run a Test Batch
print("Fetching batch...")
batch = next(iter(loader))

# Move to device
batch_gpu = {
    "seq": batch["seq"].to(device),
    "mask": batch["mask"].to(device),
    "batch_of": {k: v.to(device) for k, v in batch["batch_of"].items()}
}

# batch_gpu = crop_batch(batch_gpu, max_len=512)

print(f"Running inference for protein IDs: {batch['id']}")
with torch.no_grad():
    # Run MiniFold!
    output = model(batch_gpu)

print("\nSuccess!")
print("Latent Pair Representation shape:", output["pair"].shape) # (B, L, L, 128)
print("Predicted Coordinates shape:", output["final_atom_positions"].shape) # (B, L, 37, 3)

Using device: mps
Loading checkpoint from /Users/moustholmes/mini_latent_pd/checkpoints/minifold_cache/minifold_12L.ckpt...


NameError: name 'REPO_ID' is not defined

In [10]:
for key, value in batch.items():
    print(f"{key}: {value.shape if isinstance(value, torch.Tensor) else type(value)}")

id: <class 'list'>
seq: torch.Size([2, 883])
mask: torch.Size([2, 883])
batch_of: <class 'dict'>
gt_coords: torch.Size([2, 883, 3])


1. The Input Batch

These are the tensors fed into the model.

- id: ['1ABC', '2XYZ']

A list of the protein identifiers for tracking which file corresponds to which output.
- seq: (2, 883)

What it is: The ESM token indices for the amino acid sequence.

Details: Integers representing amino acids. The padding value (usually 1) is used for any protein shorter than 883 residues.
- mask: (2, 883)

What it is: A boolean mask indicating where real data exists versus padding.

        Values: 1 (True) for real residues, 0 (False) for padded regions. You must use this to mask your loss function during training, or the model will try to fold the padding zeros.

- batch_of: <dict>

        What it is: Pre-computed geometric features needed by the structure module (e.g., rigid body frame definitions for each amino acid type).

        Key contents: aatype (amino acid identity), rigid_group_default_frame (local frames for side chains).

- gt_coords: (2, 883, 3)

        What it is: Your Ground Truth coordinates from mdCATH.

        Note: Since the last dimension is 3, this contains only one atom per residue (likely Alpha Carbon, CA). You will use this to calculate the loss against the model's predicted CA atoms.

2. The Model Output

These are the tensors returned by model(batch).
A. The Latent Spaces 

    pair: (2, 883, 883, 128)

        What it is: The Geometry-Aware Latent Space (sz​).

        Significance: This is the output of the "Folding Trunk" (the MiniFormer). It represents the relationship between every pair of residues (L×L).

        For your project: This is the tensor you should encode/diffuse/modify. It contains the rich evolutionary and structural constraints before they are realized into 3D coordinates.

    single: (2, 883, 1024)

        What it is: The Single Latent Representation (ss​).

        Significance: A projection of the ESM embeddings mixed with structural info. It serves as the primary input to the Structure Module to generate the backbone frames.

B. The Structural Outputs (The Decoder)

These are generated by the StructureModule from the latent spaces.

    final_atom_positions: (2, 883, 37, 3)

        What it is: The predicted 3D coordinates for all heavy atoms.

        Dimensions:

            37: The standard mapping for up to 37 heavy atoms (N, CA, C, O, CB, ...).

            3: X, Y, Z coordinates (in Angstroms).

        Note: To calculate loss against your gt_coords (CA only), you need to extract the CA atom (index 1) from this tensor: pred_ca = output['final_atom_positions'][:, :, 1, :].

    final_atom_mask: (2, 883, 37)

        What it is: Tells you which of the 37 atoms actually exist for a given amino acid (e.g., Glycine has no Beta Carbon, so its index would be 0 here).

    final_affine_tensor: (2, 883, 4, 4)

        What it is: The predicted rotation and translation matrices (frames) for every residue's backbone. The Structure Module predicts these frames first, then places the atoms into them.

C. Probabilistic Outputs & Confidence

    preds: (2, 883, 883, 64)

        What it is: The Distogram Logits.

        Details: It predicts the probability of the distance between two residues falling into one of 64 bins (e.g., bin 0 = 2-4 Angstroms, bin 63 = >22 Angstroms).

    plddt: (2, 883)

        What it is: The per-residue confidence score (0-100). High values mean the model is confident in that region's structure.

    lddt_logits: (2, 883, 50)

        What it is: The raw logits used to calculate the plddt score.

    sm: <dict>

        What it is: Internal debug state of the Structure Module (e.g., side-chain torsion angles). You can usually ignore this unless you are debugging specific side-chain rotations.

In [8]:
for key, value in output.items():
    print(f"{key}: {value.shape if isinstance(value, torch.Tensor) else type(value)}")

preds: torch.Size([2, 883, 883, 64])
pair: torch.Size([2, 883, 883, 128])
single: torch.Size([2, 883, 1024])
sm: <class 'dict'>
final_atom_positions: torch.Size([2, 883, 37, 3])
final_atom_mask: torch.Size([2, 883, 37])
final_affine_tensor: torch.Size([2, 883, 4, 4])
lddt_logits: torch.Size([2, 883, 50])
plddt: torch.Size([2, 883])


In [19]:
batch['batch_of']

{'aatype': tensor([[[12],
          [12],
          [12],
          ...,
          [19],
          [19],
          [19]],
 
         [[12],
          [12],
          [12],
          ...,
          [19],
          [19],
          [19]]]),
 'seq_mask': tensor([[[1.],
          [1.],
          [1.],
          ...,
          [1.],
          [1.],
          [1.]],
 
         [[1.],
          [1.],
          [1.],
          ...,
          [1.],
          [1.],
          [1.]]]),
 'residx_atom37_to_atom14': tensor([[[[0],
           [1],
           [2],
           ...,
           [0],
           [0],
           [0]],
 
          [[0],
           [1],
           [2],
           ...,
           [0],
           [0],
           [0]],
 
          [[0],
           [1],
           [2],
           ...,
           [0],
           [0],
           [0]],
 
          ...,
 
          [[0],
           [1],
           [2],
           ...,
           [0],
           [0],
           [0]],
 
          [[0],
  

In [12]:
output['final_atom_positions'][:, :, 1, :]

tensor([[[-96.3977,   5.6547,  -6.8404],
         [-95.3969,   4.2691,  -7.1769],
         [-94.0278,   5.1603,  -7.7622],
         ...,
         [ 32.4375,   2.9705,   4.2263],
         [ 32.9936,  -0.4333,   3.6020],
         [ 33.3603,   0.5361,   0.8055]],

        [[-22.7918,   4.1711,   1.4760],
         [-23.7921,   2.8411,   3.4759],
         [-22.2749,   1.5842,   1.8703],
         ...,
         [  0.2235,  -0.8644,   0.6441],
         [  0.7852,  -1.4505,   1.4023],
         [  1.0849,  -1.3420,  -0.4275]]], device='mps:0')

In [18]:
batch['gt_coords'].to(device)-output['final_atom_positions'][:, :, 1, :]

tensor([[[ 69.3777,  47.4753,  42.9504],
         [ 68.8169,  48.2709,  42.4469],
         [ 66.2578,  48.5597,  43.3122],
         ...,
         [-87.0375,  58.2095,  30.0537],
         [-86.8236,  60.1133,  31.2680],
         [-86.4503,  60.6439,  34.6745]],

        [[ -4.5482,  57.4589,  48.6540],
         [ -3.9879,  58.8489,  45.6341],
         [ -4.5651,  61.0158,  48.4797],
         ...,
         [-21.0835,  45.5444,  63.6059],
         [-20.2252,  45.0405,  63.1377],
         [-22.1349,  44.2520,  64.9275]]], device='mps:0')

In [4]:
import torch
import torch.nn.functional as F
from pathlib import Path
from mini_latent_pd.models.minifold.model.model import MiniFoldModel
from mini_latent_pd.models.minifold.data.config import model_config
from mini_latent_pd.models.minifold.data.of_data import of_inference
from mini_latent_pd.models.minifold.utils.residue_constants import restype_order_with_x_inverse
from esm.pretrained import load_model_and_alphabet

def load_minifold_model(cache_dir="./minifold_cache", model_size="48L", device="cuda"):
    """
    Loads the MiniFold model, alphabet, and config for interactive use.
    """
    # 1. Paths
    cache = Path(cache_dir).expanduser()
    checkpoint_path = cache / f"minifold_{model_size}.ckpt"
    
    if not checkpoint_path.exists():
        raise FileNotFoundError(f"Checkpoint not found at {checkpoint_path}. Run predict.py once to download it.")

    print(f"Loading checkpoint from {checkpoint_path}...")
    
    # 2. Load Checkpoint & Hyperparams
    ckpt = torch.load(checkpoint_path, map_location="cpu")
    hparams = ckpt["hyper_parameters"]
    
    # 3. Initialize Model
    # We create the config required for feature generation
    config_of = model_config(
        "initial_training",
        train=False,
        low_prec=False,
        long_sequence_inference=False,
    )
    
    model = MiniFoldModel(
        esm_model_name=hparams["esm_model_name"],
        num_blocks=hparams["num_blocks"],
        no_bins=hparams["no_bins"],
        config_of=config_of,
        use_structure_module=True,
        kernels=False, # Important: False for Mac/MPS compatibility
    )

    # 4. Load State Dict
    state_dict = ckpt["state_dict"]
    # Clean keys as done in predict.py
    state_dict = {k: v for k, v in state_dict.items() if "boundaries" not in k}
    state_dict = {k: v for k, v in state_dict.items() if "mid_points" not in k}
    state_dict = {k.replace("_orig_mod.", ""): v for k, v in state_dict.items()}
    state_dict = {k.replace("model.", ""): v for k, v in state_dict.items()}
    
    model.load_state_dict(state_dict, strict=False)
    model.to(device)
    model.eval()
    
    # 5. Load Alphabet (needed for tokenizing sequences)
    _, alphabet = load_model_and_alphabet(hparams["esm_model_name"])
    
    return model, alphabet, config_of.data

In [5]:
from mini_latent_pd.data import MDCathToMinifoldDataset

In [6]:
from torch.utils.data import DataLoader
from mini_latent_pd.models.minifold.data import data_transforms

# 1. Define the transform to handle symmetric side-chains (Atom14 format)
def minifold_preprocess(feature_dict):
    # First create atom14 masks (atom14_atom_exists, residx_atom14_to_atom37, etc.)
    # This is required before make_atom14_positions can run.
    feature_dict = data_transforms.make_atom14_masks(feature_dict)
    
    # This transform generates 'atom14_alt_gt_positions' and 'atom14_alt_gt_exists'
    # which are strictly necessary for the OpenFold/Minifold loss functions.
    feature_dict = data_transforms.make_atom14_positions(feature_dict)
    
    # You might also want to add make_seq_mask, pseudo_beta, etc. depending 
    # on which specific Minifold modules you are fine-tuning.
    feature_dict = data_transforms.make_seq_mask(feature_dict)
    return feature_dict

# 2. Create Dataset
# Keys are 'temperature/replica/frame_index'
# Temperatures: 320, 348, 379, 413, 450
# Replicas: 0, 1, 2, 3, 4
# Frame indices: 0..439 (440 frames per replica)
md_keys = ['320/0/0', '320/0/100', '348/1/0'] 
dataset = MDCathToMinifoldDataset(
    hdf5_path='/Users/moustholmes/mini_latent_pd/data/mdcath/data/mdcath_dataset_1a0aA00.h5', 
    keys=md_keys,
    transform=minifold_preprocess
)

# 3. Handle Batching (Collation)
# Because proteins have different lengths, you cannot use the default PyTorch 
# collate_fn without padding. You can use a custom padding collate_fn, or 
# batch size = 1 if you are using gradient accumulation.
dataloader = DataLoader(
    dataset, 
    batch_size=1, # Adjust and use padding collate_fn if > 1
    shuffle=True
)

# Test iteration
for batch in dataloader:
    print("Aatype shape:", batch['aatype'].shape)                     # [1, num_res]
    print("Atom37 positions:", batch['all_atom_positions'].shape)     # [1, num_res, 37, 3]
    print("Atom14 alt exists:", batch['atom14_alt_gt_exists'].shape)  # [1, num_res, 14]
    break

Aatype shape: torch.Size([1, 63])
Atom37 positions: torch.Size([1, 63, 37, 3])
Atom14 alt exists: torch.Size([1, 63, 14])


/Users/moustholmes/mini_latent_pd/src/mini_latent_pd/models/minifold/utils/tensor_utils.py:86: UserWarning: Using a non-tuple sequence for multidimensional indexing is deprecated and will be changed in pytorch 2.9; use x[tuple(seq)] instead of x[seq]. In pytorch 2.9 this will be interpreted as tensor index, x[torch.tensor(seq)], which will result either in an error or a different result (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/autograd/python_variable_indexing.cpp:312.)
  return data[ranges]


---
# Phase 1 Prototype: Full-Atom mdCATH → MiniFold Pipeline

**Goal**: Download a small subset of mdCATH, parse full-atom coords into atom37 format, build the batch that MiniFold expects, and verify the forward pass works end-to-end.

In [1]:
## Step 1: Download a small subset of mdCATH locally
# We download a few small proteins to prototype with.
# Each protein is a separate HDF5 file (~50-200MB each).

from pathlib import Path
from huggingface_hub import hf_hub_download, list_repo_files

REPO_ID = "compsciencelab/mdCATH"
LOCAL_DIR = Path("../data/mdcath")
LOCAL_DIR.mkdir(parents=True, exist_ok=True)

# List all available files to find small proteins
all_files = list_repo_files(repo_id=REPO_ID, repo_type="dataset")
h5_files = sorted([f for f in all_files if f.endswith(".h5") and "mdcath_dataset_" in f])
print(f"Total mdCATH HDF5 files available: {len(h5_files)}")
print(f"First 10: {h5_files[:10]}")


Total mdCATH HDF5 files available: 5398
First 10: ['data/mdcath_dataset_12asA00.h5', 'data/mdcath_dataset_153lA00.h5', 'data/mdcath_dataset_16pkA02.h5', 'data/mdcath_dataset_1a02F00.h5', 'data/mdcath_dataset_1a05A00.h5', 'data/mdcath_dataset_1a0aA00.h5', 'data/mdcath_dataset_1a0hA01.h5', 'data/mdcath_dataset_1a0rP01.h5', 'data/mdcath_dataset_1a0sP00.h5', 'data/mdcath_dataset_1a15A00.h5']


In [14]:
## Step 2: Download small proteins for prototyping
# We inspect HDF5 files to find proteins with L ≤ 300 residues,
# since ESM attention is O(L^2) and blows up memory on laptop for long proteins.

import h5py
import random

MAX_SEQ_LEN = 300  # Max residues for laptop prototyping
N_TARGET = 3       # How many small proteins we want

# First, check what we already have locally
local_h5_files = sorted(LOCAL_DIR.glob("**/*.h5"))
small_local = []
for f_path in local_h5_files:
    with h5py.File(f_path, "r") as f:
        pid = list(f.keys())[0]
        n_res = len(f[pid]["resname"][:])
        print(f"  Local: {f_path.name} — {pid} L={n_res} {'✓' if n_res <= MAX_SEQ_LEN else '✗ too long'}")
        if n_res <= MAX_SEQ_LEN:
            small_local.append(f_path)

# Download more if we don't have enough small proteins
if len(small_local) < N_TARGET:
    print(f"\nNeed {N_TARGET - len(small_local)} more small proteins. Scanning remote files...")
    # Try downloading files until we have enough small ones
    for fname in h5_files:
        if len(small_local) >= N_TARGET:
            break
        
        # Check if already downloaded
        local_name = fname.split("/")[-1] if "/" in fname else fname
        if any(local_name == f.name for f in local_h5_files):
            continue
        
        print(f"  Downloading {fname}...")
        hf_hub_download(
            repo_id=REPO_ID,
            filename=fname,
            local_dir=str(LOCAL_DIR),
            repo_type="dataset",
        )
        
        # Check size
        dl_path = LOCAL_DIR / "data" / local_name if "/" in fname else LOCAL_DIR / local_name
        if not dl_path.exists():
            dl_path = list(LOCAL_DIR.glob(f"**/{local_name}"))[0]
        
        with h5py.File(dl_path, "r") as f:
            pid = list(f.keys())[0]
            n_res = len(f[pid]["resname"][:])
            print(f"    {pid}: L={n_res} {'✓ keeping' if n_res <= MAX_SEQ_LEN else '✗ too long (but keeping file)'}")
            if n_res <= MAX_SEQ_LEN:
                small_local.append(dl_path)

# List final state
local_h5_files = sorted(LOCAL_DIR.glob("**/*.h5"))
print(f"\nLocal files: {len(local_h5_files)} total")
for f_path in local_h5_files:
    with h5py.File(f_path, "r") as f:
        pid = list(f.keys())[0]
        n_res = len(f[pid]["resname"][:])
        size_mb = f_path.stat().st_size / 1e6
        print(f"  {f_path.name} — {pid} L={n_res} ({size_mb:.0f} MB) {'✓' if n_res <= MAX_SEQ_LEN else '✗'}")

print(f"\nSmall proteins (L≤{MAX_SEQ_LEN}): {len(small_local)}")

  Local: mdcath_dataset_12asA00.h5 — 12asA00 L=5091 ✗ too long
  Local: mdcath_dataset_153lA00.h5 — 153lA00 L=2855 ✗ too long
  Local: mdcath_dataset_16pkA02.h5 — 16pkA02 L=3181 ✗ too long

Need 3 more small proteins. Scanning remote files...


data/mdcath_dataset_1a02F00.h5:   0%|          | 0.00/279M [00:00<?, ?B/s]

    1a02F00: L=916 ✗ too long (but keeping file)


data/mdcath_dataset_1a05A00.h5:   0%|          | 0.00/1.60G [00:00<?, ?B/s]

    1a05A00: L=5442 ✗ too long (but keeping file)


data/mdcath_dataset_1a0aA00.h5:   0%|          | 0.00/300M [00:00<?, ?B/s]

    1a0aA00: L=1010 ✗ too long (but keeping file)


data/mdcath_dataset_1a0hA01.h5:   0%|          | 0.00/398M [00:00<?, ?B/s]

    1a0hA01: L=1338 ✗ too long (but keeping file)


data/mdcath_dataset_1a0rP01.h5:   0%|          | 0.00/594M [00:00<?, ?B/s]

    1a0rP01: L=2008 ✗ too long (but keeping file)


data/mdcath_dataset_1a0sP00.h5:   0%|          | 0.00/1.82G [00:00<?, ?B/s]

    1a0sP00: L=6200 ✗ too long (but keeping file)


data/mdcath_dataset_1a15A00.h5:   0%|          | 0.00/332M [00:00<?, ?B/s]

    1a15A00: L=1117 ✗ too long (but keeping file)


data/mdcath_dataset_1a1zA00.h5:   0%|          | 0.00/400M [00:00<?, ?B/s]

    1a1zA00: L=1357 ✗ too long (but keeping file)


data/mdcath_dataset_1a2nA02.h5:   0%|          | 0.00/931M [00:00<?, ?B/s]

    1a2nA02: L=3141 ✗ too long (but keeping file)


data/mdcath_dataset_1a39A00.h5:   0%|          | 0.00/1.80G [00:00<?, ?B/s]

    1a39A00: L=6106 ✗ too long (but keeping file)


data/mdcath_dataset_1a3dA00.h5:   0%|          | 0.00/526M [00:00<?, ?B/s]

    1a3dA00: L=1765 ✗ too long (but keeping file)


data/mdcath_dataset_1a3oA00.h5:   0%|          | 0.00/635M [00:00<?, ?B/s]

    1a3oA00: L=2138 ✗ too long (but keeping file)
    1a48A01: L=1824 ✗ too long (but keeping file)


data/mdcath_dataset_1a5cA00.h5:   0%|          | 0.00/1.55G [00:00<?, ?B/s]

    1a5cA00: L=5270 ✗ too long (but keeping file)


data/mdcath_dataset_1a66A00.h5:   0%|          | 0.00/837M [00:00<?, ?B/s]

    1a66A00: L=2843 ✗ too long (but keeping file)


KeyboardInterrupt: 

In [3]:
## Step 3: Inspect HDF5 structure of one file
# Understand the exact schema before writing the dataloader

def inspect_h5(path, max_depth=4):
    """Recursively print HDF5 structure."""
    def _visit(group, prefix=""):
        for key in group.keys():
            item = group[key]
            if isinstance(item, h5py.Group):
                print(f"{prefix}{key}/")
                if len(prefix) < max_depth * 2:
                    _visit(item, prefix + "  ")
            elif isinstance(item, h5py.Dataset):
                print(f"{prefix}{key}: shape={item.shape}, dtype={item.dtype}")
            else:
                print(f"{prefix}{key}: {type(item)}")
    
    with h5py.File(path, "r") as f:
        _visit(f)

# Inspect the first downloaded file
test_file = local_h5_files[0]
print(f"Inspecting: {test_file.name}\n")
inspect_h5(test_file)


Inspecting: mdcath_dataset_12asA00.h5

12asA00/
  320/
    0/
      box: shape=(3, 3), dtype=float32
      coords: shape=(440, 5091, 3), dtype=float32
      dssp: shape=(440, 327), dtype=object
      forces: shape=(440, 5091, 3), dtype=float32
      gyrationRadius: shape=(440,), dtype=float64
      rmsd: shape=(440,), dtype=float32
      rmsf: shape=(327,), dtype=float32
    1/
      box: shape=(3, 3), dtype=float32
      coords: shape=(440, 5091, 3), dtype=float32
      dssp: shape=(440, 327), dtype=object
      forces: shape=(440, 5091, 3), dtype=float32
      gyrationRadius: shape=(440,), dtype=float64
      rmsd: shape=(440,), dtype=float32
      rmsf: shape=(327,), dtype=float32
    2/
      box: shape=(3, 3), dtype=float32
      coords: shape=(440, 5091, 3), dtype=float32
      dssp: shape=(440, 327), dtype=object
      forces: shape=(440, 5091, 3), dtype=float32
      gyrationRadius: shape=(440,), dtype=float64
      rmsd: shape=(440,), dtype=float32
      rmsf: shape=(327,), dt

In [4]:
## Step 4: Parse pdbProteinAtoms to build flat→atom37 mapping
# This is the core of the full-atom dataloader.
# We parse the PDB text embedded in the HDF5 to map each flat atom index
# to its (residue_index, atom37_slot) position.

import numpy as np
from mini_latent_pd.models.minifold.utils.residue_constants import (
    atom_order,
    atom_types,
    restype_3to1,
    restype_order_with_x,
)

def parse_pdb_atoms(pdb_text: str):
    """
    Parse pdbProteinAtoms text from mdCATH HDF5.
    Returns:
        atom_names: list of atom name strings (e.g. ['N', 'CA', 'C', ...])
        residue_indices: list of residue index per atom (0-based)
        residue_names: list of 3-letter residue names per atom  
    """
    lines = pdb_text.strip().split("\n")
    atom_names = []
    residue_indices = []
    residue_names = []
    
    for line in lines:
        if not (line.startswith("ATOM") or line.startswith("HETATM")):
            continue
        # PDB fixed-width format:
        # cols 12-16: atom name
        # cols 17-20: residue name  
        # cols 22-26: residue sequence number
        atom_name = line[12:16].strip()
        res_name = line[17:20].strip()
        res_seq = int(line[22:26].strip())
        
        atom_names.append(atom_name)
        residue_names.append(res_name)
        residue_indices.append(res_seq)
    
    return atom_names, residue_indices, residue_names


def build_atom37_mapping(pdb_text: str):
    """
    Build the mapping from flat atom index → (residue_idx, atom37_slot).
    
    Returns:
        flat_to_res: np.array of shape (n_flat_atoms,) - residue index for each flat atom
        flat_to_atom37: np.array of shape (n_flat_atoms,) - atom37 slot for each flat atom
        valid_mask: np.array of shape (n_flat_atoms,) - True if atom is a recognized atom37 type
        n_residues: int - total number of residues
        aatype: np.array of shape (n_residues,) - restype index for each residue
        atom37_mask: np.array of shape (n_residues, 37) - which atom37 slots are present
    """
    atom_names, residue_indices, residue_names = parse_pdb_atoms(pdb_text)
    
    # Re-index residues to 0-based contiguous
    unique_res_ids = []
    seen = set()
    for r in residue_indices:
        if r not in seen:
            unique_res_ids.append(r)
            seen.add(r)
    res_id_to_idx = {r: i for i, r in enumerate(unique_res_ids)}
    n_residues = len(unique_res_ids)
    
    flat_to_res = np.zeros(len(atom_names), dtype=np.int64)
    flat_to_atom37 = np.zeros(len(atom_names), dtype=np.int64)
    valid_mask = np.zeros(len(atom_names), dtype=bool)
    atom37_mask = np.zeros((n_residues, 37), dtype=np.float32)
    
    # Determine aatype per residue (first occurrence)
    aatype = np.full(n_residues, restype_order_with_x["X"], dtype=np.int64)  # default to X
    res_idx_to_resname = {}
    
    for i, (aname, ridx, rname) in enumerate(zip(atom_names, residue_indices, residue_names)):
        res_i = res_id_to_idx[ridx]
        flat_to_res[i] = res_i
        
        # Set aatype for this residue
        if res_i not in res_idx_to_resname:
            res_idx_to_resname[res_i] = rname
            one_letter = restype_3to1.get(rname, "X")
            aatype[res_i] = restype_order_with_x.get(one_letter, restype_order_with_x["X"])
        
        # Map atom name to atom37 slot
        if aname in atom_order:
            slot = atom_order[aname]
            flat_to_atom37[i] = slot
            valid_mask[i] = True
            atom37_mask[res_i, slot] = 1.0
        # else: skip (e.g. hydrogens, OXT might not be populated)
    
    return flat_to_res, flat_to_atom37, valid_mask, n_residues, aatype, atom37_mask


# Test on the first protein
with h5py.File(test_file, "r") as f:
    protein_id = list(f.keys())[0]
    pdb_text = f[protein_id]["pdbProteinAtoms"][()].decode("utf-8")
    
    # Also grab resname for verification  
    resnames = [r.decode() for r in f[protein_id]["resname"][:]]
    
    # Check what temps/replicas are available
    temps = [k for k in f[protein_id].keys() if k.isdigit()]
    print(f"Protein: {protein_id}")
    print(f"Temperatures: {temps}")
    print(f"Resnames: {resnames[:5]}... (total: {len(resnames)})")

flat_to_res, flat_to_atom37, valid_mask, n_res, aatype, atom37_mask = build_atom37_mapping(pdb_text)

print(f"\nParsed mapping:")
print(f"  Total flat atoms: {len(flat_to_res)}")
print(f"  Valid atom37 atoms: {valid_mask.sum()}")
print(f"  Skipped atoms (H, etc): {(~valid_mask).sum()}")
print(f"  Num residues: {n_res}")
print(f"  aatype shape: {aatype.shape}")
print(f"  atom37_mask shape: {atom37_mask.shape}")
print(f"  Avg atoms per residue: {atom37_mask.sum() / n_res:.1f}")

# Verify sequence matches
from mini_latent_pd.models.minifold.utils.residue_constants import restype_order_with_x_inverse
parsed_seq = "".join([restype_order_with_x_inverse[a] for a in aatype])
print(f"\nParsed sequence: {parsed_seq}")
print(f"From resnames:   {''.join([restype_3to1.get(r, 'X') for r in resnames])}")


Protein: 12asA00
Temperatures: ['320', '348', '379', '413', '450']
Resnames: ['ALA', 'ALA', 'ALA', 'ALA', 'ALA']... (total: 5091)

Parsed mapping:
  Total flat atoms: 5091
  Valid atom37 atoms: 2558
  Skipped atoms (H, etc): 2533
  Num residues: 327
  aatype shape: (327,)
  atom37_mask shape: (327, 37)
  Avg atoms per residue: 7.8

Parsed sequence: AYIAKQRQISFVKSXFSRQLEERLGLIEVQAPILSRVGDGTQDNLSGAEKAVQVKVKALPDAQFEVVXSLAKWKRQTLGQXDFSAGEGLYTXMKALRPDEDRLSPLXSVYVDQWDWERVMGDGERQFSTLKSTVEAIWAGIKATEAAVSEEFGLAPFLPDQIXFVXSQELLSRYPDLDAKGRERAIAKDLGAVFLVGIGGKLSDGXRXDVRAPDYDDWSTPSELGXAGLNGDILVWNPVLEDAFELSSMGIRVDADTLKXQLALTGDEDRLELEWXQALLRGEMPQTIGGGIGQSRLTMLLLQLPXIGQVQAGVWPAAVRESVPSLL
From resnames:   AAAAAAAAAAAAAAAAYYYYYYYYYYYYYYYYYYYYYIIIIIIIIIIIIIIIIIIIAAAAAAAAAAKKKKKKKKKKKKKKKKKKKKKKQQQQQQQQQQQQQQQQQRRRRRRRRRRRRRRRRRRRRRRRRQQQQQQQQQQQQQQQQQIIIIIIIIIIIIIIIIIIISSSSSSSSSSSFFFFFFFFFFFFFFFFFFFFVVVVVVVVVVVVVVVVKKKKKKKKKKKKKKKKKKKKKKSSSSSSSSSSSXXXXXXXXXXXXXXXXXFFFFFFFFFFFFFFFFFFFFSSSSSSSSSSSRRRRRRRRRRR

In [3]:
## Step 5: Convert flat coords → atom37 format
# Use the mapping to scatter flat coords into (L, 37, 3) tensor

def flat_coords_to_atom37(coords_flat, flat_to_res, flat_to_atom37, valid_mask, n_residues):
    """
    Convert flat coordinates (n_atoms, 3) → atom37 format (n_residues, 37, 3).
    
    Args:
        coords_flat: (n_atoms, 3) in picometers from mdCATH
        flat_to_res: (n_atoms,) residue index per atom
        flat_to_atom37: (n_atoms,) atom37 slot per atom
        valid_mask: (n_atoms,) which atoms have a valid atom37 mapping
        n_residues: total residue count
    
    Returns:
        atom37_coords: (n_residues, 37, 3) in Angstroms
    """
    atom37_coords = np.zeros((n_residues, 37, 3), dtype=np.float32)
    
    # Only scatter valid atoms (skip hydrogens etc)
    valid_idx = np.where(valid_mask)[0]
    res_idx = flat_to_res[valid_idx]
    a37_idx = flat_to_atom37[valid_idx]
    
    # Convert pm → Å
    atom37_coords[res_idx, a37_idx] = coords_flat[valid_idx] / 10.0
    
    return atom37_coords


# Test: load one frame and convert
with h5py.File(test_file, "r") as f:
    temp = temps[0]
    repl = list(f[protein_id][temp].keys())[0]
    
    # coords dataset: (n_frames, n_all_atoms, 3) in pm
    coords_dset = f[protein_id][temp][repl]["coords"]
    print(f"Coords dataset shape: {coords_dset.shape} (frames, atoms, xyz)")
    print(f"Dtype: {coords_dset.dtype}")
    
    # Grab frame 0
    frame0 = coords_dset[0]  # (n_atoms, 3)
    print(f"Frame 0 shape: {frame0.shape}")

# Convert to atom37
atom37_frame = flat_coords_to_atom37(frame0, flat_to_res, flat_to_atom37, valid_mask, n_res)
print(f"\natom37 coords shape: {atom37_frame.shape}  (L={n_res}, 37 atom types, xyz)")

# Sanity check: extract CA coords (index 1 in atom37)
ca_coords = atom37_frame[:, 1, :]  # (L, 3)
print(f"CA coords shape: {ca_coords.shape}")
print(f"CA coords range (Å): x=[{ca_coords[:, 0].min():.1f}, {ca_coords[:, 0].max():.1f}], "
      f"y=[{ca_coords[:, 1].min():.1f}, {ca_coords[:, 1].max():.1f}], "
      f"z=[{ca_coords[:, 2].min():.1f}, {ca_coords[:, 2].max():.1f}]")

# Verify against the ca_coords dataset if available
with h5py.File(test_file, "r") as f:
    if "ca_coords" in f[protein_id][temp][repl]:
        ca_ref = f[protein_id][temp][repl]["ca_coords"][0] / 10.0  # pm → Å
        diff = np.abs(ca_coords - ca_ref)
        print(f"\nCA verification vs ca_coords dataset:")
        print(f"  Max abs diff: {diff.max():.6f} Å")
        print(f"  Mean abs diff: {diff.mean():.6f} Å")
        assert diff.max() < 0.01, "CA coords don't match!"
        print(f"  ✓ CA coords match!")
    else:
        print("\nNo ca_coords dataset for cross-check")


NameError: name 'h5py' is not defined

In [7]:
## Step 6: Build a prototype mdCATH Dataset (full-atom, local files) — using MDCathToMinifoldDataset
# Now uses the unified MDCathToMinifoldDataset in frame-pair mode.
# Lazy-loads from HDF5 on each __getitem__ — no full dataset in memory.

from pathlib import Path

LOCAL_DIR = Path("../data/mdcath")

print("Loading dataset...")
ds = MDCathToMinifoldDataset(data_dir=LOCAL_DIR, samples_per_epoch=100, max_seq_len=300)
print(f"\nDataset has {len(ds.proteins)} proteins")

if len(ds.proteins) == 0:
    print("\n⚠️  All downloaded proteins are too long! Need to download smaller ones.")
    print("Re-run Step 2 with more files (N_SUBSET=10+) to find proteins with L≤300.")
else:
    sample = ds[0]
    print(f"\nSample:")
    print(f"  id: {sample['id']}")
    print(f"  x0 shape: {sample['x0'].shape}")
    print(f"  xt shape: {sample['xt'].shape}")
    print(f"  lag: {sample['lag']} ns")
    print(f"  temp: {sample['temp']} K")
    print(f"  aatype shape: {sample['aatype'].shape}")
    print(f"  atom37_mask shape: {sample['atom37_mask'].shape}")
    print(f"  sequence: {sample['sequence'][:30]}...")

Loading dataset...
  Indexed 153lA00: L=185, 11140 frames, seq=RTDCYGNVNRIDTTGASCKT...
  Indexed 16pkA02: L=208, 11110 frames, seq=YFAKVLGNPPRPLVAIVGGA...
  Indexed 1a02F00: L=53, 10980 frames, seq=RRIRRERNKMAAAKSRNRRR...
  Indexed 1a0aA00: L=63, 10990 frames, seq=MKRESXKXAEQARRNRLAVA...
  Indexed 1a0hA01: L=89, 11030 frames, seq=SPLLETCVPDRGREYRGRLA...
  Indexed 1a0rP01: L=129, 11010 frames, seq=CMQDMXQKLSFGPRYGFVYE...
  Indexed 1a15A00: L=67, 11090 frames, seq=KPVSLSYRCPCRFFESXVAR...
  Indexed 1a1zA00: L=83, 11040 frames, seq=MDPFLVLLXSVSSSLSSSEL...
  Indexed 1a2nA02: L=208, 11060 frames, seq=AKNAALPILFAALLAEEPVE...
  Indexed 1a3dA00: L=119, 11070 frames, seq=NLYQFKNMIKCTVPSRSWWD...
  Indexed 1a3oA00: L=141, 11070 frames, seq=VLSPADKTNVKAAWGKVGAX...
  Indexed 1a48A01: L=111, 11040 frames, seq=SITKTELDGILPLVARGKVR...
  Indexed 1a66A00: L=178, 11010 frames, seq=MKDWQLPSXSGPYELRIEVQ...

Dataset has 13 proteins

Sample:
  id: 1a0aA00
  x0 shape: torch.Size([63, 37, 3])
  xt shape: torch.

In [12]:
## Step 6b: Pre-compute and save ESM embeddings (s_s, s_z) for all local proteins
# These are sequence-only (no structural info), so one embedding per protein.
# Saves {domain_id}.pt with {"s_s": (L, 1024), "s_z": (L, L, 128)}

from mini_latent_pd.data.compute_embeddings import compute_esm_embeddings

EMBEDDING_DIR = Path("../data/embeddings")

# model is already loaded from cell 8 (load_minifold_model)
# 2. Load Model
# Ensure you point cache_dir to where predict.py downloaded weights
device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
model, alphabet, config = load_minifold_model(cache_dir="/Users/moustholmes/mini_latent_pd/checkpoints/minifold_cache", model_size="12L", device=device)

compute_esm_embeddings(
    model=model,
    data_dir=LOCAL_DIR,
    output_dir=EMBEDDING_DIR,
    max_seq_len=900,
    device=str(device),
)

# Verify: load one back and inspect
import os
pt_files = sorted(EMBEDDING_DIR.glob("*.pt"))
print(f"\n--- Saved {len(pt_files)} embedding files ---")
for pt in pt_files:
    emb = torch.load(pt, map_location="cpu", weights_only=True)
    size_mb = os.path.getsize(pt) / 1e6
    print(f"  {pt.name}: s_s={tuple(emb['s_s'].shape)}, s_z={tuple(emb['s_z'].shape)}, size={size_mb:.1f}MB")

Loading checkpoint from /Users/moustholmes/mini_latent_pd/checkpoints/minifold_cache/minifold_12L.ckpt...
Found 18 HDF5 files in ../data/mdcath
  Saved 12asA00: s_s=(327, 1024), s_z=(327, 327, 128)
  Saved 153lA00: s_s=(185, 1024), s_z=(185, 185, 128)
  Saved 16pkA02: s_s=(208, 1024), s_z=(208, 208, 128)
  Saved 1a02F00: s_s=(53, 1024), s_z=(53, 53, 128)
  Saved 1a05A00: s_s=(357, 1024), s_z=(357, 357, 128)
  Saved 1a0aA00: s_s=(63, 1024), s_z=(63, 63, 128)
  Saved 1a0hA01: s_s=(89, 1024), s_z=(89, 89, 128)
  Saved 1a0rP01: s_s=(129, 1024), s_z=(129, 129, 128)
  Saved 1a0sP00: s_s=(413, 1024), s_z=(413, 413, 128)
  Saved 1a15A00: s_s=(67, 1024), s_z=(67, 67, 128)
  Saved 1a1zA00: s_s=(83, 1024), s_z=(83, 83, 128)
  Saved 1a2nA02: s_s=(208, 1024), s_z=(208, 208, 128)
  Saved 1a39A00: s_s=(401, 1024), s_z=(401, 401, 128)
  Saved 1a3dA00: s_s=(119, 1024), s_z=(119, 119, 128)
  Saved 1a3oA00: s_s=(141, 1024), s_z=(141, 141, 128)
  Saved 1a48A01: s_s=(111, 1024), s_z=(111, 111, 128)
  Saved

In [13]:
## Step 6c: Test dataset with pre-computed embeddings
# Pass embedding_dir to MDCathToMinifoldDataset — embeddings are lazy-loaded per __getitem__

ds_with_emb = MDCathToMinifoldDataset(
    data_dir=LOCAL_DIR,
    embedding_dir=EMBEDDING_DIR,
    samples_per_epoch=100,
    max_seq_len=300,
)

print(f"Proteins indexed: {len(ds_with_emb.proteins)}")
for p in ds_with_emb.proteins:
    print(f"  {p['id']}: emb={'✓' if p['embedding_path'] else '✗'}")

if len(ds_with_emb.proteins) > 0:
    sample = ds_with_emb[0]
    print(f"\nSample keys: {list(sample.keys())}")
    print(f"  x0: {sample['x0'].shape}")
    print(f"  xt: {sample['xt'].shape}")
    if "s_s" in sample:
        print(f"  s_s: {sample['s_s'].shape}  (single repr)")
        print(f"  s_z: {sample['s_z'].shape}  (pair repr)")
        print(f"  s_s stats: mean={sample['s_s'].mean():.4f}, std={sample['s_s'].std():.4f}")
        print(f"  s_z stats: mean={sample['s_z'].mean():.4f}, std={sample['s_z'].std():.4f}")
    else:
        print("  ⚠️ No embeddings found — run Step 6b first")

  Indexed 153lA00: L=185, 11140 frames, seq=RTDCYGNVNRIDTTGASCKT..., emb=✓
  Indexed 16pkA02: L=208, 11110 frames, seq=YFAKVLGNPPRPLVAIVGGA..., emb=✓
  Indexed 1a02F00: L=53, 10980 frames, seq=RRIRRERNKMAAAKSRNRRR..., emb=✓
  Indexed 1a0aA00: L=63, 10990 frames, seq=MKRESXKXAEQARRNRLAVA..., emb=✓
  Indexed 1a0hA01: L=89, 11030 frames, seq=SPLLETCVPDRGREYRGRLA..., emb=✓
  Indexed 1a0rP01: L=129, 11010 frames, seq=CMQDMXQKLSFGPRYGFVYE..., emb=✓
  Indexed 1a15A00: L=67, 11090 frames, seq=KPVSLSYRCPCRFFESXVAR..., emb=✓
  Indexed 1a1zA00: L=83, 11040 frames, seq=MDPFLVLLXSVSSSLSSSEL..., emb=✓
  Indexed 1a2nA02: L=208, 11060 frames, seq=AKNAALPILFAALLAEEPVE..., emb=✓
  Indexed 1a3dA00: L=119, 11070 frames, seq=NLYQFKNMIKCTVPSRSWWD..., emb=✓
  Indexed 1a3oA00: L=141, 11070 frames, seq=VLSPADKTNVKAAWGKVGAX..., emb=✓
  Indexed 1a48A01: L=111, 11040 frames, seq=SITKTELDGILPLVARGKVR..., emb=✓
  Indexed 1a66A00: L=178, 11010 frames, seq=MKDWQLPSXSGPYELRIEVQ..., emb=✓
Proteins indexed: 13
  153lA00

In [13]:
## Step 7: Build collator that produces the batch format MiniFold expects
# MiniFold's forward() expects:
#   batch["seq"]:      (B, L) ESM token indices
#   batch["mask"]:     (B, L) boolean mask
#   batch["batch_of"]: dict with:
#     - "aatype":                     (B, L, num_recycle=1)
#     - "seq_mask":                   (B, L, 1)
#     - "residx_atom37_to_atom14":    (B, L, 37)
#     - "atom37_atom_exists":         (B, L, 37)

from mini_latent_pd.models.minifold.data.config import model_config
from mini_latent_pd.models.minifold.data.of_data import of_inference
from mini_latent_pd.models.minifold.utils.residue_constants import restype_order_with_x_inverse
from esm.pretrained import load_model_and_alphabet

class MDCATHMiniFoldCollator:
    """
    Collates mdCATH atom37 samples into batches that MiniFold can process.
    Generates the OpenFold features from sequences and handles padding.
    """
    def __init__(self, alphabet, of_config):
        self.alphabet = alphabet
        self.of_config = of_config
    
    def __call__(self, batch):
        seq_tokens_list = []
        masks_list = []
        batch_of_list = []
        x0_list = []
        xt_list = []
        ids = []
        lags = []
        temps = []
        
        for item in batch:
            seq_str = item["sequence"]
            ids.append(item["id"])
            lags.append(item["lag"])
            temps.append(item["temp"])
            
            # Generate OpenFold features from sequence
            of_feats = of_inference(seq_str, "predict", self.of_config)
            
            # Reconstruct clean sequence
            clean_seq = "".join(
                [restype_order_with_x_inverse[x.item()] for x in of_feats["aatype"]]
            )[:of_feats["seq_length"]]
            
            # ESM tokenization
            encoded_seq = torch.tensor(self.alphabet.encode(clean_seq), dtype=torch.long)
            
            seq_tokens_list.append(encoded_seq)
            masks_list.append(of_feats["seq_mask"][:, 0].bool())
            
            # Only keep the features MiniFold needs
            relevant_keys = {"aatype", "seq_mask", "residx_atom37_to_atom14", "atom37_atom_exists"}
            batch_of_list.append({k: v for k, v in of_feats.items() if k in relevant_keys})
            
            x0_list.append(item["x0"])
            xt_list.append(item["xt"])
        
        # Padding
        max_len = max(len(s) for s in seq_tokens_list)
        
        # Pad ESM tokens
        padded_seqs = torch.stack([
            F.pad(s, (0, max_len - len(s)), value=1) for s in seq_tokens_list
        ])
        
        # Pad masks
        padded_masks = torch.stack([
            F.pad(m, (0, max_len - len(m)), value=0) for m in masks_list
        ])
        
        # Pad OpenFold features (dim 0 is L)
        batched_of = {}
        if batch_of_list:
            for k in batch_of_list[0].keys():
                feats = [d[k] for d in batch_of_list]
                padded = []
                for feat in feats:
                    pad_size = max_len - feat.shape[0]
                    # Pad on the first dim (L), keep others
                    pad_spec = [0] * 2 * (len(feat.shape) - 1) + [0, pad_size]
                    padded.append(F.pad(feat, pad_spec, value=0))
                batched_of[k] = torch.stack(padded)
        
        # Pad coords: (B, L, 37, 3)
        padded_x0 = torch.stack([
            F.pad(x, (0, 0, 0, 0, 0, max_len - x.shape[0]), value=0.0)
            for x in x0_list
        ])
        padded_xt = torch.stack([
            F.pad(x, (0, 0, 0, 0, 0, max_len - x.shape[0]), value=0.0)
            for x in xt_list
        ])
        
        return {
            "id": ids,
            "seq": padded_seqs,
            "mask": padded_masks,
            "batch_of": batched_of,
            "x0": padded_x0,        # (B, L, 37, 3) ground truth frame 0
            "xt": padded_xt,         # (B, L, 37, 3) ground truth frame t
            "lag": torch.tensor(lags, dtype=torch.long),
            "temp": torch.tensor(temps, dtype=torch.long),
        }


print("Collator class defined ✓")


Collator class defined ✓


In [14]:
## Step 8: End-to-end test — mdCATH → MiniFold forward pass
# Load model, create dataloader, run forward pass, compare predictions to GT

from torch.utils.data import DataLoader

# 1. Setup device
device = torch.device("mps" if torch.backends.mps.is_available() else "cpu")
print(f"Device: {device}")

# 2. Load MiniFold (use 12L for laptop prototyping)
model, alphabet, of_data_config = load_minifold_model(
    cache_dir="./minifold_cache", model_size="12L", device=device
)

# 3. Create collator and dataloader
collator = MDCATHMiniFoldCollator(alphabet, of_data_config)
loader = DataLoader(ds, batch_size=1, collate_fn=collator, shuffle=False)

# 4. Get a batch
batch = next(iter(loader))
print(f"\nBatch contents:")
for k, v in batch.items():
    if isinstance(v, torch.Tensor):
        print(f"  {k}: {v.shape} {v.dtype}")
    elif isinstance(v, dict):
        print(f"  {k}:")
        for kk, vv in v.items():
            print(f"    {kk}: {vv.shape} {vv.dtype}")
    else:
        print(f"  {k}: {type(v)}")


Device: mps
Loading checkpoint from minifold_cache/minifold_12L.ckpt...

Batch contents:
  id: <class 'list'>
  seq: torch.Size([1, 178]) torch.int64
  mask: torch.Size([1, 178]) torch.bool
  batch_of:
    aatype: torch.Size([1, 178, 1]) torch.int64
    seq_mask: torch.Size([1, 178, 1]) torch.float32
    residx_atom37_to_atom14: torch.Size([1, 178, 37, 1]) torch.int64
    atom37_atom_exists: torch.Size([1, 178, 37, 1]) torch.float32
  x0: torch.Size([1, 178, 37, 3]) torch.float32
  xt: torch.Size([1, 178, 37, 3]) torch.float32
  lag: torch.Size([1]) torch.int64
  temp: torch.Size([1]) torch.int64


/Users/moustholmes/mini_latent_pd/src/mini_latent_pd/models/minifold/data/data_transforms.py:1231: UserWarning: Using a non-tuple sequence for multidimensional indexing is deprecated and will be changed in pytorch 2.9; use x[tuple(seq)] instead of x[seq]. In pytorch 2.9 this will be interpreted as tensor index, x[torch.tensor(seq)], which will result either in an error or a different result (Triggered internally at /Users/runner/work/pytorch/pytorch/pytorch/torch/csrc/autograd/python_variable_indexing.cpp:312.)
  protein[k] = v[slices]


In [15]:
## Step 9: Run MiniFold forward and compare outputs to GT coords

MAX_LEN = 900  # Safety crop — ESM attention is O(L^2), keep ≤300 on laptop

def crop_minifold_batch(batch, max_len):
    """Crop all tensors in a MiniFold batch to max_len residues."""
    L = batch["seq"].shape[1]
    if L <= max_len:
        return batch
    print(f"⚠️  Cropping batch from L={L} to {max_len}")
    cropped = {
        "seq": batch["seq"][:, :max_len],
        "mask": batch["mask"][:, :max_len],
        "batch_of": {},
    }
    for k, v in batch["batch_of"].items():
        cropped["batch_of"][k] = v[:, :max_len, ...] if v.shape[1] == L else v
    return cropped

# Move batch to device and crop if needed
batch_gpu = {
    "seq": batch["seq"].to(device),
    "mask": batch["mask"].to(device),
    "batch_of": {k: v.to(device) for k, v in batch["batch_of"].items()},
}
# batch_gpu = crop_minifold_batch(batch_gpu, MAX_LEN)

L = int(batch_gpu["mask"][0].sum().item())
print(f"Running MiniFold on protein: {batch['id']} (L={L})")
with torch.no_grad():
    output = model(batch_gpu)

print(f"\nModel outputs:")
for k, v in output.items():
    if isinstance(v, torch.Tensor):
        print(f"  {k}: {v.shape}")
    elif isinstance(v, dict):
        print(f"  {k}: <dict with {len(v)} keys>")
    else:
        print(f"  {k}: {type(v)}")

# Extract predictions
pred_atom37 = output["final_atom_positions"].cpu()  # (B, L, 37, 3)
pred_ca = pred_atom37[:, :, 1, :]                    # (B, L, 3) — CA is index 1

# Ground truth CA from our atom37 coords (crop to match)
gt_x0_cropped = batch["x0"][:, :pred_ca.shape[1], :, :]
gt_ca_x0 = gt_x0_cropped[:, :, 1, :]

# Compute RMSD (masked)
pred_ca_valid = pred_ca[0, :L]
gt_ca_x0_valid = gt_ca_x0[0, :L]

# Simple RMSD (no alignment — just to check scale/reasonability)
rmsd_vs_x0 = torch.sqrt(((pred_ca_valid - gt_ca_x0_valid) ** 2).sum(-1).mean())

print(f"\nResults for {batch['id'][0]} (L={L}):")
print(f"  MiniFold pred CA vs GT frame 0 — RMSD: {rmsd_vs_x0:.2f} Å (unaligned)")
print(f"  (Large RMSD is expected — MiniFold predicts from sequence, not from MD coords)")

# The pair representation z — this is what we'll use as the latent space
pair_repr = output["pair"]  # (B, L, L, 128)
print(f"\n  Pair representation z: {pair_repr.shape}  ← this is the latent space")
print(f"  z stats: mean={pair_repr.mean():.4f}, std={pair_repr.std():.4f}")
print(f"  z memory: {pair_repr.element_size() * pair_repr.nelement() / 1e6:.1f} MB")

Running MiniFold on protein: ['1a66A00'] (L=178)

Model outputs:
  preds: torch.Size([1, 178, 178, 64])
  pair: torch.Size([1, 178, 178, 128])
  single: torch.Size([1, 178, 1024])
  sm: <dict with 7 keys>
  final_atom_positions: torch.Size([1, 178, 37, 3])
  final_atom_mask: torch.Size([1, 178, 37])
  final_affine_tensor: torch.Size([1, 178, 4, 4])
  lddt_logits: torch.Size([1, 178, 50])
  plddt: torch.Size([1, 178])

Results for 1a66A00 (L=178):
  MiniFold pred CA vs GT frame 0 — RMSD: 96.51 Å (unaligned)
  (Large RMSD is expected — MiniFold predicts from sequence, not from MD coords)

  Pair representation z: torch.Size([1, 178, 178, 128])  ← this is the latent space
  z stats: mean=5.8816, std=46.1227
  z memory: 16.2 MB


In [16]:
## Step 10: Prototype the Geometry Encoder (coords → distogram → z)
# This tests the autoencoder encoding path from the plan:
#   1. coords → distogram (via Cβ distances)  
#   2. z = s_z_static + recycle(distogram)
# Where s_z_static comes from the ESM embeddings (sequence-level, static)

# Extract the pretrained components from MiniFold
fold_trunk = model.fold
s_z_static_proj = fold_trunk.projection       # Linear(3*128, 128)
recycle_linear = fold_trunk.recycle            # Linear(64, 128) — distogram → z
seq_to_pair = fold_trunk.seq_to_pair           # SequenceToPair(1024, 64, 128)
pos_embed = fold_trunk.positional_embedding    # RelativePosition(32, 128)

print("Extracted FoldingTrunk components:")
print(f"  projection: {s_z_static_proj}")
print(f"  recycle: {recycle_linear}")
print(f"  seq_to_pair: {seq_to_pair}")
print(f"  pos_embed: {pos_embed}")

# Run the ESM part manually to get s_s and s_z
with torch.no_grad():
    idx = model.lm.num_layers
    lm_out = model.lm(
        batch_gpu["seq"], repr_layers=[idx], need_head_weights=True
    )
    s_s = model.fc_s(lm_out["representations"][idx])  # (B, L, 1024)
    B, N = s_s.shape[:2]
    s_z = lm_out["attentions"].view(B, N, N, -1)
    s_z = model.fc_z(s_z)                              # (B, L, L, 128)

print(f"\nESM embeddings:")
print(f"  s_s (single repr): {s_s.shape}")
print(f"  s_z (pair repr from attention): {s_z.shape}")

# Now build s_z_static = proj(cat[s_z, seq_to_pair(s_s), rel_pos(residx, mask)])
mask = batch_gpu["mask"]
pair_mask = mask[:, None, :] * mask[:, :, None]
residx = torch.arange(N, device=device).unsqueeze(0).expand(B, -1)

with torch.no_grad():
    s_z_static = torch.cat([
        s_z,
        seq_to_pair(s_s),
        pos_embed(residx, mask=pair_mask),
    ], dim=-1)
    s_z_static = s_z_static_proj(s_z_static)

print(f"  s_z_static: {s_z_static.shape}")

# Now test: coords → distogram → recycle → z
from mini_latent_pd.models.minifold.utils.residue_constants import atom_order

def pseudo_beta_from_atom37(atom37_coords, aatype):
    """
    Extract Cβ coords (or Cα for Glycine) from atom37 format.
    Args:
        atom37_coords: (B, L, 37, 3)
        aatype: (B, L) residue type indices
    Returns:
        pseudo_beta: (B, L, 3)
    """
    # CB is at index 3, CA is at index 1
    ca = atom37_coords[..., 1, :]   # (B, L, 3)
    cb = atom37_coords[..., 3, :]   # (B, L, 3)
    
    # Glycine (index 7 in restype_order) has no CB → use CA
    is_gly = (aatype == 7)  # G=7 in restypes order: A R N D C Q E G...
    pseudo_beta = torch.where(is_gly.unsqueeze(-1), ca, cb)
    return pseudo_beta


def coords_to_distogram(pseudo_beta, min_bin=2.3125, max_bin=21.6875, no_bins=64):
    """
    Compute one-hot distogram from pseudo-beta positions.
    Same binning as MiniFold's distogram_loss.
    
    Args:
        pseudo_beta: (B, L, 3) Cβ/Cα positions in Angstroms
    Returns:
        distogram: (B, L, L, no_bins) one-hot encoded distances
    """
    # Pairwise squared distances
    diffs = pseudo_beta[:, :, None, :] - pseudo_beta[:, None, :, :]  # (B, L, L, 3)
    dist_sq = (diffs ** 2).sum(-1)  # (B, L, L)
    
    # Bin boundaries (squared, as in loss.py)
    sq_breaks = torch.linspace(min_bin**2, max_bin**2, no_bins - 1, device=pseudo_beta.device)
    
    # Bin assignment
    bin_idx = torch.bucketize(dist_sq, sq_breaks)  # (B, L, L), values in [0, no_bins-1]
    distogram = F.one_hot(bin_idx, no_bins).float()  # (B, L, L, no_bins)
    
    return distogram


# Apply to GT coords
gt_x0_gpu = batch["x0"].to(device)
aatype_gpu = batch["batch_of"]["aatype"][:, :, 0].long().to(device)  # remove recycle dim

pseudo_cb = pseudo_beta_from_atom37(gt_x0_gpu, aatype_gpu)
disto = coords_to_distogram(pseudo_cb)

print(f"\nDistogram from GT coords:")
print(f"  pseudo_beta: {pseudo_cb.shape}")
print(f"  distogram: {disto.shape}")

# Encode: z = s_z_static + recycle(distogram)
with torch.no_grad():
    z_encoded = s_z_static + recycle_linear(disto.to(device))

print(f"\nEncoded z: {z_encoded.shape}")
print(f"  z stats: mean={z_encoded.mean():.4f}, std={z_encoded.std():.4f}")
print(f"\n✓ Geometry encoder works! This z encodes both sequence AND structure info")


Extracted FoldingTrunk components:
  projection: Linear(in_features=384, out_features=128, bias=True)
  recycle: Linear(in_features=64, out_features=128, bias=True)
  seq_to_pair: SequenceToPair(
  (layernorm): LayerNorm((1024,), eps=1e-05, elementwise_affine=True)
  (proj): Linear(in_features=1024, out_features=128, bias=True)
  (o_proj): Linear(in_features=128, out_features=128, bias=True)
)
  pos_embed: RelativePosition(
  (embedding): Embedding(66, 128)
)

ESM embeddings:
  s_s (single repr): torch.Size([1, 178, 1024])
  s_z (pair repr from attention): torch.Size([1, 178, 178, 128])
  s_z_static: torch.Size([1, 178, 178, 128])

Distogram from GT coords:
  pseudo_beta: torch.Size([1, 178, 3])
  distogram: torch.Size([1, 178, 178, 64])

Encoded z: torch.Size([1, 178, 178, 128])
  z stats: mean=-0.0538, std=9.9250

✓ Geometry encoder works! This z encodes both sequence AND structure info


In [17]:
## Step 11: Prototype the Decoder (z → structure)
# Test path: z_encoded → PairToSequence → StructureModule → atom14 → atom37

sz_project = model.sz_project      # PairToSequence(c_z=128, c_s=1024)
structure_module = model.structure_module

# Decode: z → single repr → structure
with torch.no_grad():
    pair_mask = (batch_gpu["mask"][:, None, :] * batch_gpu["mask"][:, :, None])
    
    # PairToSequence: (B, L, L, 128) + (B, L, 1024) → (B, L, 1024)
    s_single = sz_project(z_encoded, s_s, pair_mask)
    print(f"Single repr from decoder: {s_single.shape}")
    
    # StructureModule: needs (s, z, aatype, mask)
    feats = {k: v.to(device) for k, v in batch["batch_of"].items()}
    feats_no_recycle = {k: v[..., 0] if v.dim() > 2 else v for k, v in feats.items()}
    
    sm_out = structure_module(
        s=s_single,
        z=z_encoded,
        aatype=feats_no_recycle["aatype"],
        mask=feats_no_recycle["seq_mask"].to(dtype=s_single.dtype),
    )
    
    # atom14 → atom37
    from mini_latent_pd.models.minifold.utils.feats import atom14_to_atom37
    pred_positions = atom14_to_atom37(sm_out["positions"][-1], feats_no_recycle)
    
print(f"StructureModule output positions: {sm_out['positions'][-1].shape}")  # (B, L, 14, 3)
print(f"Decoded atom37 positions: {pred_positions.shape}")                    # (B, L, 37, 3)

# Compare decoded positions to GT
L = batch_gpu["mask"][0].sum().item()
pred_ca = pred_positions[0, :L, 1, :].cpu()   # CA
gt_ca = batch["x0"][0, :L, 1, :]

rmsd = torch.sqrt(((pred_ca - gt_ca) ** 2).sum(-1).mean())
print(f"\nAutoencoder roundtrip (encode GT → decode):")
print(f"  CA RMSD vs GT: {rmsd:.2f} Å")
print(f"  (This tests encode→decode with pretrained weights, no fine-tuning)")


Single repr from decoder: torch.Size([1, 178, 1024])
StructureModule output positions: torch.Size([1, 178, 14, 3])
Decoded atom37 positions: torch.Size([1, 178, 37, 3])

Autoencoder roundtrip (encode GT → decode):
  CA RMSD vs GT: 94.84 Å
  (This tests encode→decode with pretrained weights, no fine-tuning)


In [18]:
## Step 12: Prototype the velocity field (latent dynamics)
# Test: given two frames (x0, xt), encode both → compute target velocity → 
# test MiniFormer prediction

miniformer = model.fold.miniformer

with torch.no_grad():
    # Encode x0
    pb_x0 = pseudo_beta_from_atom37(batch["x0"].to(device), aatype_gpu)
    d_x0 = coords_to_distogram(pb_x0)
    z_x0 = s_z_static + recycle_linear(d_x0.to(device))
    
    # Encode xt
    pb_xt = pseudo_beta_from_atom37(batch["xt"].to(device), aatype_gpu)
    d_xt = coords_to_distogram(pb_xt)
    z_xt = s_z_static + recycle_linear(d_xt.to(device))
    
    # Target velocity
    v_target = z_xt - z_x0  # (B, L, L, 128)
    
    # MiniFormer prediction (pretrained, no scale factor yet)
    pair_mask = (batch_gpu["mask"][:, None, :] * batch_gpu["mask"][:, :, None]).to(z_x0)
    z_pred = miniformer(z_x0, pair_mask)  # (B, L, L, 128)
    
    # v_pred = MiniFormer(z_source) * scale_factor - z_source
    # With scale_factor=1.0 initially:
    v_pred = z_pred - z_x0
    
    # Compute loss (MSE on velocity, masked)
    L_val = int(batch_gpu["mask"][0].sum().item())
    mask_2d = pair_mask[0, :L_val, :L_val]
    
    v_target_valid = v_target[0, :L_val, :L_val]  # (L, L, 128)
    v_pred_valid = v_pred[0, :L_val, :L_val]
    
    mse = ((v_pred_valid - v_target_valid) ** 2).mean()

print(f"Latent dynamics prototype:")
print(f"  z_x0: {z_x0.shape}")
print(f"  z_xt: {z_xt.shape}")
print(f"  lag: {batch['lag']} ns")
print(f"  v_target norm: {v_target_valid.norm():.4f}")
print(f"  v_pred norm: {v_pred_valid.norm():.4f}")
print(f"  velocity MSE: {mse:.6f}")
print(f"\n✓ Full pipeline works: mdCATH → atom37 → encode → velocity → loss")


Latent dynamics prototype:
  z_x0: torch.Size([1, 178, 178, 128])
  z_xt: torch.Size([1, 178, 178, 128])
  lag: tensor([125]) ns
  v_target norm: 1049.7891
  v_pred norm: 72806.1719
  velocity MSE: 1307.047485

✓ Full pipeline works: mdCATH → atom37 → encode → velocity → loss


## Summary: What was prototyped

| Step | What | Status |
|------|------|--------|
| 1-2 | Download mdCATH subset (3 small proteins) | ✓ |
| 3 | Inspect HDF5 schema | ✓ |
| 4 | Parse `pdbProteinAtoms` → flat→atom37 mapping | ✓ |
| 5 | Convert flat coords → atom37 `(L, 37, 3)` + verify CA against `ca_coords` | ✓ |
| 6 | `MDCATHAtom37Dataset` — loads full-atom frame pairs from local HDF5 | ✓ |
| 7 | `MDCATHMiniFoldCollator` — pads + generates OpenFold features for MiniFold | ✓ |
| 8-9 | End-to-end: mdCATH batch → MiniFold forward → compare pred vs GT | ✓ |
| 10 | Geometry Encoder: coords → distogram → `z = s_z_static + recycle(disto)` | ✓ |
| 11 | Decoder: `z → PairToSequence → StructureModule → atom37` (roundtrip test) | ✓ |
| 12 | Velocity field: encode frame pair → target velocity → MiniFormer loss | ✓ |

**Next steps**: Move prototyped code into proper modules (`src/mini_latent_pd/data/mdcath_dataset.py`, `models/structure_autoencoder.py`, etc.) as per Phase 1-2 of the plan.